Upcast writing and View enhancements

f1ames opened the pull request:
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Other: Introduced `createDocumentFragment()`, `createElement()` and `createText()` methods in `UpcastWriter`. The `View.change()` method now returns result of its callback. Closes #1549.

---

### Additional information

See #1549 for details.

f1ames commented:
There is related PR (change in docs) in `clipboard` plugin - https://github.com/ckeditor/ckeditor5-clipboard/pull/59.

Reinmar commented on src/view/documentfragment.js:
```diff
@@ -20,6 +20,9 @@ export default class DocumentFragment {
 	 * Creates new DocumentFragment instance.
 	 *
+	 * **Note:** Constructor of this class shouldn't be used directly in the code. Use the
+	 * {@link module:engine/view/upcastwriter~UpcastWriter#createDocumentFragment} method instead.
```
```suggestion
	 * {@link module:engine/view/upcastwriter~UpcastWriter#createDocumentFragment `UpcastWriter#createDocumentFragment()`}
	 * method instead.
```

Reinmar replied:
Please repeat that change in other places as well. Without the additional link text, such a link generates as a useless `createDocumentFragment`. Without clicking it you don't know where to find it. Plus, it does not render with `()` so without reading the text it does not look like a method.

Reinmar commented on src/view/upcastwriter.js:
```diff
@@ -17,6 +19,49 @@ import { isPlainObject } from 'lodash-es';
  * see {@link module:engine/view/downcastwriter~DowncastWriter writer}.
  */
 export default class UpcastWriter {
+	/**
```
Missing blank line.

Reinmar commented on src/view/upcastwriter.js:
```diff
@@ -17,6 +19,49 @@ import { isPlainObject } from 'lodash-es';
  */
 export default class UpcastWriter {
+	/**
+	 * Creates new {@link module:engine/view/documentfragment~DocumentFragment}.
```
```suggestion
	 * Creates a new {@link module:engine/view/documentfragment~DocumentFragment}.
```

We don't need to specify the link's text in this case because we'd just repeat the default one.

Reinmar commented on src/view/upcastwriter.js:
```diff
@@ -17,6 +19,49 @@ import { isPlainObject } from 'lodash-es';
+	 * @see module:engine/view/documentfragment~DocumentFragment#constructor
+	 * @param {module:engine/view/node~Node|Iterable.<module:engine/view/node~Node>} [children]
+	 * List of nodes to be inserted into created document fragment.
+	 * @returns {module:engine/view/documentfragment~DocumentFragment} Created document fragment.
```
```suggestion
	 * @returns {module:engine/view/documentfragment~DocumentFragment} The created document fragment.
```

Reinmar commented on src/view/upcastwriter.js:
```diff
@@ -17,6 +19,49 @@ import { isPlainObject } from 'lodash-es';
+	 *
+	 * Attributes can be passed in various formats:
+	 *
+	 *		new Element( 'div', { 'class': 'editor', 'contentEditable': 'true' } ); // object
```
This code sample is incorrect. You shouldn't use `new Element()`. Plus, you don't need to wrap those key names in `''`.

Reinmar commented on src/view/upcastwriter.js:
```diff
@@ -17,6 +19,49 @@ import { isPlainObject } from 'lodash-es';
+	}
+
+	/**
+	 * Creates new {@link module:engine/view/element~Element}.
```
```suggestion
	 * Creates a new {@link module:engine/view/element~Element} instance.
```

Reinmar commented on src/view/upcastwriter.js:
```diff
@@ -17,6 +19,49 @@ import { isPlainObject } from 'lodash-es';
+	}
+
+	/**
+	 * Creates new {@link module:engine/view/text~Text}.
```
```suggestion
	 * Creates a new {@link module:engine/view/text~Text} instance.
```

Reinmar commented on src/view/upcastwriter.js:
```diff
@@ -17,6 +19,49 @@ import { isPlainObject } from 'lodash-es';
+	 *
+	 * @see module:engine/view/text~Text#constructor
+	 * @param {String} data Text
+	 * @returns {module:engine/view/text~Text} Created text.
```
```suggestion
	 * @returns {module:engine/view/text~Text} The created text node.
```

Reinmar commented on src/view/upcastwriter.js:
```diff
@@ -17,6 +19,49 @@ import { isPlainObject } from 'lodash-es';
+	 * Creates new {@link module:engine/view/text~Text}.
+	 *
+	 * @see module:engine/view/text~Text#constructor
+	 * @param {String} data Text
```
```suggestion
	 * @param {String} data The text's data.
```

Reinmar commented on src/view/upcastwriter.js:
```diff
@@ -17,6 +19,49 @@ import { isPlainObject } from 'lodash-es';
+	/**
+	 * Creates new {@link module:engine/view/text~Text}.
+	 *
+	 * @see module:engine/view/text~Text#constructor
```
Don't link to protected things from public things. If there's any useful documentation in that protected API, move it here.

Reinmar commented on src/view/view.js:
```diff
@@ -308,16 +308,19 @@ export default class View {
-	 *		view.change( writer => {
-	 *			writer.insert( position1, writer.createText( 'foo' ) );
+	 *		const text = view.change( writer => {
+	 *			const newText = writer.createText( 'foo' );
```
Missing blank line.

Reinmar commented on src/view/view.js:
```diff
@@ -308,16 +308,19 @@ export default class View {
 	 * It makes sure that after all changes are made view is rendered to DOM. It prevents situations when DOM is updated
 	 * when view state is not yet correct. It allows to nest calls one inside another and still perform single rendering
-	 * after all changes are applied.
+	 * after all changes are applied. It also returns the result of its callback.
```
```suggestion
	 * after all changes are applied. It also returns the return value of its callback.
```

Reinmar commented on src/view/element.js:
```diff
@@ -53,8 +53,9 @@ export default class Element extends Node {
-	 * {@link module:engine/view/downcastwriter~DowncastWriter#createUIElement} for UI element instead.
+	 * {@link module:engine/view/downcastwriter~DowncastWriter#createEmptyElement} for empty element,
+	 * {@link module:engine/view/downcastwriter~DowncastWriter#createUIElement} for UI element instead or
+	 * {@link module:engine/view/upcastwriter~UpcastWriter#createElement} for general element creation.
```
I'd remove this documentation from here (and move it to the class docs if that information is missing there). Let's keep the maximum in the class definition. This method is protected so not many people will see it. 

Please also review other element/text/df classes for the same issue.

Reinmar commented:
A control question – did you check for `new Element/DF/Text()` use in at least API docs? I'm sure there's a lot of that in tests, but that we can leave (it's also correct cause constructors are protected, not private).

f1ames commented:
When it comes to https://github.com/ckeditor/ckeditor5-engine/pull/1579/commits/b0aa93b01d1214b4faa9fc1f8f94ee706a611dbf, I have changed all API samples which used protected constructors (e.g. `new Element()`) to use `DowncastWriter`.

It could be also changed to use `UpcastWriter`, however as `UpcastWriter` doesn't provide methods to create specialized elements (like `downcastWriter.createAttributeElement()`) some samples would require using both downcast and upcast writer which may look confusing.
